Site Overview Retrofit: Backport the site overview v2 to the current hosting dashboard

arthur791004 opened the pull request:
<!--
Link a related Github/Linear issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword. Consider using "part of" instead.
-->

Part of DOTDASH-145

## Proposed Changes

* Introduce the `dashboard/v2/backport/site-overview` flag to control the backport
* Backport the site overview v2 to the current hosting dashboard
* Update the backport URL prefix to enable seamless tab navigation (e.g., between Overview and Settings) without triggering a full page reload
* Adjust the grid layout to prevent the content from overflowing
  * Use `Grid` for the nested cards instead of `VStack` to make the items with the same height
  * Adjust the style of the `OverviewCard` to prevent the content from overflowing
* The breakpoint of the layout
  * with site list: 1440px
  * without site list: 960px

| | Large screen | Small screen |
|-|-|-|
| With site list |<img width="1634" height="1292" alt="image" src="https://github.com/user-attachments/assets/7ceed2ea-ecdd-496b-a7c4-5a12ac05213f" />|<img width="1429" height="1293" alt="image" src="https://github.com/user-attachments/assets/840eb073-760a-4529-9f9c-1941af92f8e0" />|
| Without site list |<img width="1272" height="1293" alt="image" src="https://github.com/user-attachments/assets/41423e3b-537b-4b4d-b63c-90972475b7e4" />|<img width="949" height="1293" alt="image" src="https://github.com/user-attachments/assets/da0cbb60-044b-45be-b598-3a3c1b814ea0" />|

## Why are these changes being made?
<!--
It's easy to see what a PR does but much harder to find out why it was made,
particularly when researching old changes in history. Record an explanation of
the motivation behind this change and how it will help.
-->

*

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Go to `/sites?flags=dashboard/v2/backport/site-overview`
* Select any site
* Ensure the backport looks good

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you tested accessibility for your changes? Ensure the feature remains usable with various user agents (e.g., browsers), interfaces (e.g., keyboard navigation), and assistive technologies (e.g., screen readers) (PCYsg-S3g-p2).
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

arthur791004 committed:
Introduce dashboard/v2/backport/site-overview flag

arthur791004 committed:
Implement SiteOverview backport

fushar commented on client/sites/v2/site-overview/router.tsx:
```diff
@@ -0,0 +1,86 @@
+	component: () => <Outlet />,
+} );
+
+const siteOverviewRoute = createRoute( {
```
Do we need the preload data fetching logic here? 🤔

arthur791004 replied:
Done by https://github.com/Automattic/wp-calypso/pull/104828/commits/3ab52cd5123c5e1def5b895ff3699781e5404f2c.

fushar replied:
I just realized that I don't think it's possible. I just tested hovering the links in the site list and it does not trigger data fetch for the site's overview. I believe it's because the site list and site overview backports have different TanStack routers 🥲 

I'm not sure yet how to solve this...

arthur791004 replied:
Tried it via forwarding the loader to the `router.preloadRoute()`. See https://github.com/Automattic/wp-calypso/pull/104828/commits/836ae4e80dccf44f7c578480dcaee5dbd4ea5c5e.

I guess we can also import the specific route and override the properties as below:

```js
const siteOverviewCompatibilityRoute = createRoute( {
	...siteOverviewRoute,
	getParentRoute: () => rootRoute,
	path: '/sites/$siteSlug',
	beforeLoad: ( { cause, params: { siteSlug } } ) => {
		if ( cause !== 'enter' ) {
			return;
		}
		throw redirect( { to: `/sites/v2/${ siteSlug }`, replace: true } );
	},
} );
```

fushar replied:
> I guess we can also import the specific route 

Ah, do you mean `siteOverviewRoute` [from v2](https://github.com/Automattic/wp-calypso/blob/b5bca365d6079f5611fe154ec27abe5aa2ba6165/client/dashboard/app/router.tsx#L131)? Does that mean we can eliminate duplication here because we can just do `loader: siteOverviewRoute.loader`? If so then that's awesome 😄

arthur791004 replied:
Yes I guess we can do it but we have to export every route 😄

fushar replied:
But they are already exported no?

https://github.com/Automattic/wp-calypso/blob/b5bca365d6079f5611fe154ec27abe5aa2ba6165/client/dashboard/app/router.tsx#L628

arthur791004 replied:
Oh yes, you're right!

arthur791004 replied:
There is a type issue because the path is different 😓

fushar reviewed (commented):
The Visibility card seems to be broken; if I click it I just got a blank page:

<img width="640" height="649" alt="Image" src="https://github.com/user-attachments/assets/6ed4bb3d-ecbf-4bc6-8ea3-85e219347840" />

arthur791004 committed:
Preload queries

arthur791004 committed:
Change the backport url prefix to /sites/v2 to resolve the redirection issue between feature tabs

arthur791004 commented:
> The Visibility card seems to be broken; if I click it I just got a blank page:

Fixed by https://github.com/Automattic/wp-calypso/pull/104828/commits/d044c4fb3b8bfe5fee92e3678149d52a891d715a!

arthur791004 committed:
Preload route on hover

matt-west reviewed (approved):
Looking good!

There's a bit of a weird layout shift as you resize the window. The layout changes to one-column, then back to three-columns again for a bit after the sites list is hidden.

This should be resolved once we get rid of that sidebar sites list though. 🤞

fushar commented on client/dashboard/app/router.tsx:
```diff
@@ -253,6 +253,9 @@ const siteSettingsPHPRoute = createRoute( {
 	getParentRoute: () => siteRoute,
 	path: 'settings/database',
+	loader: async ( { params: { siteSlug } } ) => {
+		await queryClient.ensureQueryData( siteBySlugQuery( siteSlug ) );
```
Isn't this already ensured in the parent route?

arthur791004 replied:
This is to avoid the TS error for the backport 🥲

arthur791004 replied:
Or would it be better to make it as a dummy loader?

fushar replied:
What do you mean by "dummy loader"?

arthur791004 replied:
Just a function that does nothing 😄

```js
loader: () => {}
```

fushar replied:
I found that this works 🙈 🏃‍♂️ 

```ts
const databaseRoute = createRoute( {
	path: ( appRouter.siteSettingsDatabaseRoute.options as { path: string } ).path,
	loader: appRouter.siteSettingsDatabaseRoute.options.loader,
	getParentRoute: () => siteRoute,
} ).lazy( () =>
```

arthur791004 replied:
How about just duplicate the path here instead of reusing the route 🤔 I don't think the path changes very often

fushar replied:
Sure, it's just to demonstrate how to hack TS haha

arthur791004 committed:
Refactor

fushar commented on client/sites/v2/site-overview/router.tsx:
```diff
@@ -0,0 +1,70 @@
+	)
+);
+
+const siteSettingsPreloadRoute = createRoute( {
```
Could you explain what `siteSettingsPreloadRoute` and `siteSettingsSiteVisibilityPreloadRoute` are used for here? In my testing, they don't seem to be doing anything 🤔

arthur791004 replied:
When you hover on a link that points to site settings or site visibility settings, it will trigger the loader to preload the data. Since these routes have different loader, we need to have both. Maybe we can remove `siteSettingsPreloadRoute` as the site overview screen doesn't have any link pointing to the site settings page 🤔

fushar replied:
> site visibility settings

Is it for the Visibility Card? I tested and it does not preload anything when I hover the card 🤔 could you double check?

fushar replied:
Sorry I think you're right, I add some debugging and it seems it preloads correctly 😄 

And yeah `siteSettingsPreloadRoute` does not do anything as there's no link to it. Can be removed.

arthur791004 replied:
I'll remove it. Besides, I'll change to use preload route in v1 since the current way won't load the js files...

fushar commented on client/sites/v2/site-settings/layout.tsx:
```diff
@@ -5,16 +5,13 @@ import { Provider as ReduxProvider } from 'react-redux';
 
-// Serves a similar purpose to AppConfig form v2 dashboard.
-const config = {
-	basePath: '/',
```
I think I agree with @p-jackson 's [comment here](https://github.com/Automattic/wp-calypso/pull/104921#discussion_r2241136257) that it is not actually a router config, but app config instead. So it's fine to leave it here as-is, mimicking the actual dashboard app 🤔 

Is there a strong reason why we must refactor it? I see that this config is duplicated anyway in site list, settings, and overview backport code.

arthur791004 replied:
In v2, we use a single router that's dynamically created based on different app configurations. However, in v1, we have multiple routers for the backport, and there are no separate apps involved. Additionally, we need to create the router upfront because some routes need to preload paths that are registered by other routes.

Given these structural differences and the lack of configuration flexibility in v1, I don't think we need to follow the same pattern as in v2. The requirements and setup are fundamentally different, so it makes more sense to handle routing in a way that's tailored to v1.

What do you think?

fushar replied:
> Additionally, we need to create the router upfront because some routes need to preload paths that are registered by other routes.

It makes sense now... 😄

fushar commented on client/sites/v2/site-overview/controller.tsx:
```diff
@@ -0,0 +1,14 @@
+	next: () => void
+) {
+	if ( isEnabled( 'dashboard/v2/backport/site-overview' ) ) {
+		return page.redirect( `/sites/v2/${ context.params.site }` );
```
```suggestion
		return page.redirect( `/sites/${ context.params.site }` );
```

arthur791004 replied:
oops...missing this due to rebasing 🤦‍♂️

fushar reviewed (commented):
It seems the empty state of DataViews is a bit off now

<img width="602" height="157" alt="Image" src="https://github.com/user-attachments/assets/79765a3f-326b-45c8-9d0f-2caa89f85ed5" />

arthur791004 committed:
Improve the preload

fushar reviewed (commented):
When the feature flag is NOT set, I'm seeing a flashing white screen again when clicking a site from the site list. Seems a regression after you fixed it in the below PR, could we check?

- https://github.com/Automattic/wp-calypso/pull/104492

arthur791004 committed:
Fix white screen

arthur791004 commented:
> When the feature flag is NOT set, I'm seeing a flashing white screen again when clicking a site from the site list. Seems a regression after you fixed it in the below PR, could we check?

Fixed by https://github.com/Automattic/wp-calypso/pull/104828/commits/46639771276d773d2503b17f607403ae67020cc7. The reason is the `/sites/` (the end of slash) prevents the site dashboard from rendering before the redirection.

Besides, I also removed the redirection as the new URL is what we want in the future and we can render the current overview by returning the `overview(...)`.

There is still an issue that the animation seems a bit slow... 🫠

fushar reviewed (approved):
I think it's good now, we just need to fix the tests 👍
